### ISSUE TITLE
IServiceProvider is disposed in ResolveFieldContext in subscriptions

### ISSUE THREAD

**Issue by ganhammar** [graphql-dotnet/server#706]
I'm having issues with DataLoaded fields in subscriptions, the instance of `IServiceProvider` in `ResolveFieldContextAdapter.RequestServices` is disposed so I can't resolve any dependencies required to resolve that DataLoaded field. It might be something that I'm missing, the GraphQL server configuration in `Startup` looks like this:

```csharp
GraphQL.MicrosoftDI.GraphQLBuilderExtensions.AddGraphQL(services)
    .AddSubscriptionDocumentExecuter()
    .AddServer(true)
    .AddSchema<ApiSchema>()
    .AddDefaultEndpointSelectorPolicy()
    .AddSystemTextJson()
    .AddWebSockets()
    .AddDataLoader()
    .AddGraphTypes()
    .AddUserContextBuilder(x => new UserContext(x.User));
```

The schema is setup code first, the subscription field looks something like this:

```csharp
subscription.AddField(new EventStreamFieldType
{
    Name = "OnThingChange",
    Arguments = new QueryArguments(
        new QueryArgument<NonNullGraphType<IdGraphType>> { Name = "Id" }),
    Type = typeof(ThingGraphType),
    Resolver = new FuncFieldResolver<Thing>(context =>
    {
        var thing = context.Source as Thing;

        return thing;
    }),
    Subscriber = new EventStreamResolver<Thing>(context =>
    {
        var id = context.GetArgument<Guid>("Id");

        return context.RequestServices.GetRequiredService<ThingSubscription>().Stream
            .Where(x => x.Id == id);
    }),
});
```

And the DataLoaded field in question looks something like this (it blows up at the first `GetService`):

```csharp
Field<ListGraphType<ThingPartGraphType>, IEnumerable<ThingPart>>()
    .Name("ThingParts")
    .ResolveAsync(context =>
    {
        var dbContext = context.GetService<ApiDbContext>(); // <--- Exception thrown: 'System.ObjectDisposedException' in Microsoft.Extensions.DependencyInjection.dll: 'Cannot access a disposed object.'
        var accessor = context.GetService<IDataLoaderContextAccessor>();
        var loader = accessor.Context.GetOrAddCollectionBatchLoader<Guid, ThingPart>(
            "GetThingParts",
            async (things) =>
            {
               ...
            });
        return loader.LoadAsync(context.Source.Id);
    });
```

I have verified that it's actually the `IServiceProvider` instance that is disposed by trying to create a new scope (`context.RequestServices.CreateScope();`) with the same result.

Any ideas? Thanks!

**Comment by sungam3r** [comment]
@Shane32 Relates to #691 ?

> This PR creates a scope for the duration of the execution. The scope and acquired services are disposed immediately after execution completes. In this way, there are no unintended extremely-long lifetimes of scoped objects.

@ganhammar I think that such behavior is by design.

**Comment by ganhammar** [comment]
@sungam3r I see, the intention makes sense but it seems like it might be disposed to early, otherwise I have to handle graph types used by subscriptions differently?

Upon some further investigation it seems like it is disposed already in the `FuncFieldResolver` of the subscription.

**Comment by sungam3r** [comment]
Waiting for further comments from @Shane32 .

**Comment by Shane32** [comment]
@ganhammar I may not understand subscriptions fully, but it looks like you have a subscription field and a regular query field.  And it's crashing on the regular query field?  And before it hits the data loader at all?

**Comment by Shane32** [comment]
Perhaps you can post a sample graphql request that triggers the error (matching your above sample).

**Comment by ganhammar** [comment]
@Shane32 Sorry, my bad, I should have explained it better. So, I have a Subscription called `OnThingChange` which is a subscription of `ThingGraphType`. `ThingGraphType` contains a data loaded field called `ThingParts`, that field is included in the subscription graphql request. When a new instance of `Thing` is added to the stream and the resolver of the subscription is triggered the `IServiceProvider` is already disposed. It's therefore also disposed when trying to resolve the data loaded field `ThingParts`.

So, in this case I'm only talking about a subscription and the part that looks like a query is actually the data loaded field of the `ThingGraphType`.

The graphql request for the subscription looks something like this:

```graphql
subscription($id: ID!) {
  onThingChange(id: $id) {
    id
    thingParts {
      id
    }
  }
}
```

Does it help?

I can try to recreate the issue in a small sample as soon as I have the time :)

**Comment by Shane32** [comment]
Ok that makes sense.  The service provider is only active for the initial subscription field resolver.  This is by design, and also a bug: we should have provided `null` to the RequestServices field for the child fields.

We can make a change to provide a service scope to each execution, but this is complicated because of a couple reason.  Maybe v5, and perhaps it would be an optional feature.

We can also keep the service provider alive for the entire duration of the subscription.  But this is per-connection, and could be a drain in system resources, as the service provider may be alive for a very long time.  For example, Entity Framework tracks objects pulled from the database.  With a long-life service provider, eventually your RAM will fill with EF tracking so many objects.  (And EF performance will suffer also.)  So I don't think this is the right answer.

You can solve this now by adding `IServiceScopeFactory` to the constructor of `ThingGraphType` and call `_serviceScopeFactory.CreateScope()` within the field resolver to create a service scope.  Of course if you don't need scoped services you can put singleton services directly in the constructor.

To fix the bug we need to add:

```cs
resolveContext.RequestServices = null;
```

right before this line:

https://github.com/graphql-dotnet/graphql-dotnet/blob/951753aa59451083df9ea61609b81a462e0e48ff/src/GraphQL.SystemReactive/SubscriptionExecutionStrategy.cs#L105

**Comment by ganhammar** [comment]
@Shane32 Okay, makes sense, I'll follow your suggestion! Thanks! I could create a PR for the fix if it would help you guys?

**Comment by Shane32** [comment]
That would be great, thanks!

**Comment by Shane32** [comment]
This was previously patched in graphql-dotnet/graphql-dotnet#2847 but will be reverted in v5 so that better support of scoped services within subscriptions are possible.
